### PR TITLE
support for ubuntu UEFI boot

### PR DESCRIPTION
**Pull request by zhenz**

### CODE REVIEW

File: cc-snapshot [hunk 1]
```diff
@@ -308,51 +291,70 @@ if [ $DISTRO = $UBUNTU ]; then
...
+            exit 1
+        ;;
+    esac
+    CC_SNAPSHOT_CONVERTED_PATH=/tmp/snapshot.img
+    filesystem_size=$(du -d 0 -m --exclude-from=$EXCLUDE_FROM / |  awk -F" "  '{ print $1 }')
+    total_disk_size=$(($filesystem_size + 550 + 8))
+
+    guestfish -N $CC_SNAPSHOT_CONVERTED_PATH=disk:"$total_disk_size"M -- \
+    part-init /dev/sda gpt : \
+    part-add /dev/sda primary 2048 1128447 : \ # 550M for efi partition
```
Comment (diurnalist) [inline cc-snapshot]: Is 550M standard for this? Does this depend on the EFI image size?

**Review by diurnalist — APPROVED** [review]
Some of those part-add values are a bit scary to me, b/c they seem so specific. Is this approach inspired by something you've seen elsewhere?

Otherwise I trust that it works well, and I'm not sure if folks were using partition images. Do you have any idea?

**Comment by zhenz** [comment]
> Some of those part-add values are a bit scary to me, b/c they seem so specific. Is this approach inspired by something you've seen elsewhere?
> 
> Otherwise I trust that it works well, and I'm not sure if folks were using partition images. Do you have any idea?

@diurnalist I admit the numbers look scary, but I tried several things, and this is the only workable solution by far. I stick to the tool used before, which is https://libguestfs.org/guestfish.1.html. And for 550M and 8M, I followed https://github.com/openstack/diskimage-builder/blob/master/diskimage_builder/elements/block-device-efi/block-device-default.yaml. 

And for partition image, I believe the ARM image used to be a partition image (days before I joined I think). but all chameleon-supported images are whole disk images now, and I don't think partition image support in cc-snapshot is used.

**Review by msherman64 — APPROVED** [review]
Looks good to me!

File: cc-snapshot [hunk 2]
```diff
@@ -308,51 +291,70 @@ if [ $DISTRO = $UBUNTU ]; then
...
+    part-add /dev/sda primary 1128448 1144831 : \ # 8M for bios partition
+    part-add /dev/sda primary 1144832 -40 : \
+    mkfs $FS /dev/sda3 label:$LABEL : \
+    mkfs vfat /dev/sda1 : \
+    mount /dev/sda3 / : tar-in $CC_SNAPSHOT_TAR_PATH / : mount /dev/sda1 /boot/efi : \
+    sh "parted /dev/sda set 2 bios_grub on" : \
+    sh "parted /dev/sda set 1 esp on" : \
+    sh "grub-install --target=$GRUB_INSTALL_TARGET --efi-directory=/boot/efi --removable && grub-install /dev/sda && grub-mkconfig -o /boot/grub/grub.cfg"
```
Comment (msherman64) [inline cc-snapshot:355]: This looks like the "magic" that makes the grub install work now?

File: cc-snapshot [hunk 3]
```diff
@@ -308,51 +291,70 @@ if [ $DISTRO = $UBUNTU ]; then
...
+    esac
+    CC_SNAPSHOT_CONVERTED_PATH=/tmp/snapshot.img
+    filesystem_size=$(du -d 0 -m --exclude-from=$EXCLUDE_FROM / |  awk -F" "  '{ print $1 }')
+    total_disk_size=$(($filesystem_size + 550 + 8))
+
+    guestfish -N $CC_SNAPSHOT_CONVERTED_PATH=disk:"$total_disk_size"M -- \
+    part-init /dev/sda gpt : \
+    part-add /dev/sda primary 2048 1128447 : \ # 550M for efi partition
+    part-add /dev/sda primary 1128448 1144831 : \ # 8M for bios partition
+    part-add /dev/sda primary 1144832 -40 : \
```
Comment (msherman64) [inline cc-snapshot:349]: is the -40 for the backup gpt sector at the end?

Comment (zhenz) [reply cc-snapshot:349]: I'm not sure, but I can't use up the disk. so I believe it's saved for something.

File: cc-snapshot (hunk 2 above)
Comment (zhenz) [reply cc-snapshot:355]: yes :)

**Pull request merged**